src/repr/src/row.rs

```diff
@@ -427,7 +456,44 @@ fn read_untagged_bytes<'a>(data: &'a [u8], offset: &mut usize) -> &'a [u8] {
 /// This function is safe if the datum's length and contents were previously written by `push_lengthed_bytes`,
 /// and it was only written with a `String` tag if it was indeed UTF-8.
 unsafe fn read_lengthed_datum<'a>(data: &'a [u8], offset: &mut usize, tag: Tag) -> Datum<'a> {
-    let len = match tag {
+    let len = read_lengthened_datum_len(data, offset, tag);
+    let bytes = &data[*offset..(*offset + len)];
+    *offset += len;
+    match tag {
+        Tag::BytesTiny | Tag::BytesShort | Tag::BytesLong | Tag::BytesHuge => Datum::Bytes(bytes),
+        Tag::StringTiny | Tag::StringShort | Tag::StringLong | Tag::StringHuge => {
+            Datum::String(str::from_utf8_unchecked(bytes))
+        }
+        _ => unreachable!(),
+    }
+}
+
+/// Read a datum kind whose length is encoded in the row before its contents.
+///
+/// Updates `offset` to point to the first byte after the end of the read region.
+///
+/// # Safety
+///
+/// This function is safe if the datum's length and contents were previously written by `push_lengthed_bytes`,
+/// and it was only written with a `String` tag if it was indeed UTF-8.
+unsafe fn read_lengthed_datum_kind(data: &[u8], offset: &mut usize, tag: Tag) -> DatumKind {
+    let len = read_lengthened_datum_len(data, offset, tag);
+    *offset += len;
+    match tag {
+        Tag::BytesTiny | Tag::BytesShort | Tag::BytesLong | Tag::BytesHuge => DatumKind::Bytes,
+        Tag::StringTiny | Tag::StringShort | Tag::StringLong | Tag::StringHuge => DatumKind::String,
+        _ => unreachable!(),
+    }
+}
+
+/// Read the length of a datum whose length is incoded in the row before its contents.
+///
+/// # Safety
+///
+/// This function is safe if the datum's length and contents were previously written by `push_lengthed_bytes`,
+/// and it was only written with a `String` tag if it was indeed UTF-8.
+unsafe fn read_lengthened_datum_len(data: &[u8], offset: &mut usize, tag: Tag) -> usize {
```

Comment:
We use *lengthed* everywhere else, but *lengthened* here. *Lengthed* is probably not a real English word, but the meaning here is more or less clear.